### PR TITLE
#149 - Create `minos-router-graphql` plugin

### PR DESCRIPTION
**Pull request by vladyslav-fenchak**

### CODE REVIEW

File: packages/plugins/minos-router-graphql/tests/test_graphql/test_handlers.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+async def resolve_ticket(request: Request):
+    return Response(3)
+
+
+async def resolve_ticket_raises(request: Request):
+    raise ResponseException("Some error.", status=403)
+
+
+class TestGraphQlHttpRouter(unittest.IsolatedAsyncioTestCase):
```
Comment (garciparedes) [inline packages/plugins/minos-router-graphql/tests/test_graphql/test_handlers.py:48]:
```suggestion
class TestGraphQlHandler(unittest.IsolatedAsyncioTestCase):
```

File: packages/plugins/minos-router-graphql/minos/plugins/graphql/routers.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+from .builders import (
+    GraphQLSchemaBuilder,
+)
+from .handlers import (
+    GraphQlHandler,
+)
+
+
+class GraphQlHttpRouter(HttpRouter):
+    """TODO"""
```
Comment (garciparedes) [inline packages/plugins/minos-router-graphql/minos/plugins/graphql/routers.py:24]:
```suggestion
    """GraphQl Http Router class."""
```

File: packages/plugins/minos-router-graphql/tests/test_graphql/test_decorators.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+)
+from minos.plugins.graphql.decorators import (
+    GraphQlQueryEnrouteDecorator,
+)
+from tests.utils import (
+    FakeQueryService,
+)
+
+
+class TestSomething(unittest.TestCase):
```
Comment (garciparedes) [inline packages/plugins/minos-router-graphql/tests/test_graphql/test_decorators.py:22]:
```suggestion
class TestGraphQlEnrouteDecorator(unittest.TestCase):
```

File: packages/plugins/minos-router-graphql/tests/test_graphql/test_decorators.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    GraphQLString,
+)
+
+from minos.common import (
+    classname,
+)
+from minos.networks import (
+    EnrouteAnalyzer,
+)
+from minos.plugins.graphql.decorators import (
```
Comment (garciparedes) [inline packages/plugins/minos-router-graphql/tests/test_graphql/test_decorators.py:14]: Hi @vladyslav-fenchak,

You must include decorator classes to be able to import them directly from `minos.plugins.graphql` 🙂

File: packages/plugins/minos-router-graphql/tests/test_graphql/test_builders/test_schema.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    BASE_PATH,
+    user_type,
+)
+
+
+async def callback_fn(request: Request):
+    return Response("ticket #4")
+
+
+class TestSchemaBuilder(unittest.IsolatedAsyncioTestCase):
```
Comment (garciparedes) [inline packages/plugins/minos-router-graphql/tests/test_graphql/test_builders/test_schema.py:34]:
```suggestion
class TestGraphQLSchemaBuilder(unittest.IsolatedAsyncioTestCase):
```

File: packages/plugins/minos-router-graphql/minos/plugins/graphql/decorators.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+class GraphQlEnroute:
+    """GraphQl Enroute class"""
+
+    command = GraphQlCommandEnrouteDecorator
+    query = GraphQlQueryEnrouteDecorator
+
+    @classmethod
+    def register(cls):
+        """TODO"""
```
Comment (garciparedes) [inline packages/plugins/minos-router-graphql/minos/plugins/graphql/decorators.py:51]:
```suggestion
        """Register the graphql sub-enroute."""
```

File: packages/plugins/minos-router-graphql/minos/plugins/graphql/decorators.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+from minos.networks import (
+    EnrouteDecorator,
+    EnrouteDecoratorKind,
+    enroute,
+)
+
+
+class GraphQlEnrouteDecorator(EnrouteDecorator, ABC):
+    """GraphQl Enroute class"""
```
Comment (garciparedes) [inline packages/plugins/minos-router-graphql/minos/plugins/graphql/decorators.py:20]:
```suggestion
    """GraphQl Enroute Decorator class"""
```

File: packages/plugins/minos-router-graphql/minos/plugins/graphql/decorators.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        self.name = name
+        self.argument = argument
+        self.output = output
+
+    def __iter__(self) -> Iterable:
+        yield from (self.name,)
+
+
+class GraphQlCommandEnrouteDecorator(GraphQlEnrouteDecorator):
+    """GraphQl Command Enroute class"""
```
Comment (garciparedes) [inline packages/plugins/minos-router-graphql/minos/plugins/graphql/decorators.py:32]:
```suggestion
    """GraphQl Command Enroute Decorator class"""
```

File: packages/plugins/minos-router-graphql/minos/plugins/graphql/decorators.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+    KIND: Final[EnrouteDecoratorKind] = EnrouteDecoratorKind.Command
+
+
+class GraphQlQueryEnrouteDecorator(GraphQlEnrouteDecorator):
+    """GraphQl Query Enroute class"""
```
Comment (garciparedes) [inline packages/plugins/minos-router-graphql/minos/plugins/graphql/decorators.py:38]:
```suggestion
    """GraphQl Query Enroute Decorator class"""
```

File: packages/plugins/minos-router-graphql/minos/plugins/graphql/handlers.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        result = await graphql(schema=self._schema, source=source, variable_values=variables)
+
+        errors = result.errors
+        if errors is None:
+            errors = list()
+
+        status = 200
+
+        if len(errors):
+            status = 400
```
Comment (garciparedes) [inline packages/plugins/minos-router-graphql/minos/plugins/graphql/handlers.py:47]: The default `status` should be `500`, only overwritten by a `ResponseException` status 🙂

File: packages/plugins/minos-router-graphql/minos/plugins/graphql/builders/schema.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+    def _build_queries(cls, routes):
+        fields = dict()
+        for route, callback in routes.items():
+            callback = cls.adapt_callback(callback)
+            if route.KIND == EnrouteDecoratorKind.Query:
+                fields[route.name] = cls._build_field(route, callback)
+
+        result = GraphQLObjectType(
+            "Query", fields={"hello": GraphQLField(GraphQLString, resolve=lambda obj, info: "world")}
+        )
```
Comment (garciparedes) [inline packages/plugins/minos-router-graphql/minos/plugins/graphql/builders/schema.py:74]: We should try to find a different solution for this problem. 😅

Comment (vladyslav-fenchak) [reply packages/plugins/minos-router-graphql/minos/plugins/graphql/builders/schema.py:74]: I have simplified it, but it is still necessary to pass a base structure. The good thing is that this is transparent to the user.
```python
result = GraphQLObjectType(
            "Query", fields={"dummy": GraphQLString}
        )
```

https://stackoverflow.com/questions/54322029/graphqljs-query-root-type-must-be-provided

**Review by andrea-mucci — APPROVED** [review]

**Review by garciparedes — APPROVED** [review]

File: packages/plugins/minos-router-graphql/minos/plugins/graphql/builders/schema.py (hunk 11 above)
Comment (garciparedes) [reply packages/plugins/minos-router-graphql/minos/plugins/graphql/builders/schema.py]: You were right @vladyslav-fenchak , I do not understand why is that constraint but we need to follow it hahaha

**Pull request merged**